model/jpa/src/main/java/org/keycloak/connections/jpa/updater/liquibase/custom/JpaUpdate22_0_0_RemoveRhssoThemes.java

```diff
@@ -43,10 +43,11 @@ protected void generateStatementsImpl() throws CustomChangeException {
                 .addWhereParameter("rh-sso"));
         // remove account theme for realms
         statements.add(new UpdateStatement(null, null, database.correctObjectName("REALM", Table.class))
-                .addNewColumnValue("ACCOUNT_THEME", null)
-                .setWhereClause("ACCOUNT_THEME=? OR ACCOUNT_THEME=?")
+                .addNewColumnValue("ACCOUNT_THEME", "keycloak.v2")
+                .setWhereClause("ACCOUNT_THEME=? OR ACCOUNT_THEME=? OR ACCOUNT_THEME=?")
                 .addWhereParameter("rh-sso")
-                .addWhereParameter("rh-sso.v2"));
+                .addWhereParameter("rh-sso.v2")
```

Comment:
I'm a bit worried to see an existing changeset which was included in 22.0.0 already change here. Does liquibase calculate the changeset hashes based on the actual SQL generated here? 

Can you provide more details why this change was necessary - I didn't see comments in the issue of the pull request.

Reply:
AFAIK, and based on some experiments in the past with Liquibase, it calculates the hashes based on the SQL that is produced by the change.

That being said, I wonder what would happen in successive migrations - i.e. one to 22 using the old code, and then another one to 22.0.x where this change lives. I think it will fail with an error due to mismatching hashes, no?

Reply:
> I removed the use of the constant from services. This avoids a dependency, and at the same time this migration should migrate a specific value that is now current, and should not change when someone decides to change the value of the constant.

@ahus1 thank you

> I'm a bit worried to see an existing changeset which was included in 22.0.0 already change here. Does liquibase calculate the changeset hashes based on the actual SQL generated here?

According to local testing it doesn't seem so. When I start Keycloak 22.0.4 e.g. with postgres db, then stop it and then start Keycloak with this change it doesn't throw any exceptions and I can see that last `EXECUTED` changeset in the `databasechangelog` table is the on with id `22.0.0-17484`. 

> Can you provide more details why this change was necessary

The aim is to make migration consistent for both when migrating data already present in the database and when migrating a realm exported in previous versions of keycloak.

As far as I saw there was an [PR](https://github.com/keycloak/keycloak/pull/19963) which aimed to remove `rh-sso` and `rh-sso.v2` themes. And then another one which removed old account console: https://github.com/keycloak/keycloak/pull/21098/files#diff-42324cb8460623d9089f7e7620891a25778bcdc32c6fa73ab1ccc4f3c8842454R72-R74

It seems to me that desired outcome is to combine the two and this change aims to achieve it. 

It might be cleaner to introduce new changelog with this change.

Reply:
Thank you  @sguilhen. ^^

Reply:
> According to local testing it doesn't seem so. When I start Keycloak 22.0.4 e.g. with postgres db, then stop it and then start Keycloak with this change it doesn't throw any exceptions and I can see that last `EXECUTED` changeset in the `databasechangelog` table is the on with id `22.0.0-17484`.
> 
 This is is the id of the change in `jpa-changelog-22.0.0.xml`, which Liquibase has marked as `EXECUTED` when you started KC 22.0.4. I think it doesn't attempt to re-run the change when you run Keycloak with this change. Have you checked the DB to see if the new changes were applied? I would expect them not to be.

Reply:
I'm pretty sure this is an old way of handling the checksums... see [this page](https://docs.liquibase.com/commands/maintenance/calculate-checksum.html#:~:text=When%20running%20the%20calculate%2Dchecksum,been%20run%20against%20the%20database.) from Liquibase docs, where it says "When running the calculate-checksum command, the DATABASECHANGELOG table calculates an MD5 checksum for each entry based on the SQL script of the changeset. This checksum helps Liquibase detect differences between the changesets you want to deploy and the changesets that have already been run against the database."

Perhaps using the XML as the basis for the checksum was the way they did in the past, but I think it was fragile - if you add an space inside the XML, the checksum would be different although the SQL that effectively changes the DB is the same. Not to mention that Liquibase accepts other formats for the changesets, such as JSON and YAML. Regardless of format used, if a particular change produces a SQL that is exactly the same as one that was already executed, Liquibase knows it is safe to skip it.

Reply:
I did a local test and it confirmed Vlasta's finding as changing the generated SQL didn't break the checksum, only changing the XML. Let's have a short 3-people knowledge share / knowledge discovery eventually if there's interest.